pkg/recommender/recommender.go

```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
+	}
+
+	return nil, nil
+}
+
+func (r *recommender) getDatacenterRecommendations(ctx context.Context, datacenter *scyllav1.DatacenterSpec, scalingPolicy *v1alpha1.DatacenterScalingPolicy) (*v1alpha1.DatacenterRecommendations, error) {
+	var rackRecommendations []v1alpha1.RackRecommendations
+	for _, rackScalingPolicy := range scalingPolicy.RackScalingPolicies {
+		var rack scyllav1.RackSpec
+
+		found := false
+		for i := range datacenter.Racks {
+			if datacenter.Racks[i].Name == rackScalingPolicy.Name {
+				rack = datacenter.Racks[i]
+				found = true
+				break
 			}
+		}
 
-			dcRecommendations = append(dcRecommendations, v1alpha1.DataCenterRecommendations{Name: dc.Name, RackRecommendations: rackRecommendations})
+		if !found {
+			return nil, errors.New(fmt.Sprintf("rack \"%s\" not found", rackScalingPolicy.Name))
 		}
 
-		sca.Status.Recommendations = &v1alpha1.ScyllaClusterRecommendations{DataCenterRecommendations: dcRecommendations}
-		err = r.client.Status().Update(ctx, sca)
+		recommendations, err := r.getRackRecommendations(ctx, &rack, &rackScalingPolicy)
 		if err != nil {
-			r.logger.Error(ctx, "SCA status update", "error", err)
+			return nil, errors.Wrap(err, fmt.Sprintf("rack \"%s\"", rack.Name))
+		}
+		if recommendations != nil {
+			rackRecommendations = append(rackRecommendations, *recommendations)
+		}
+	}
+
+	if len(rackRecommendations) > 0 {
+		return &v1alpha1.DatacenterRecommendations{Name: datacenter.Name, RackRecommendations: rackRecommendations}, nil
+	}
+
+	return nil, nil
+}
+
+func (r *recommender) getRackRecommendations(ctx context.Context, rack *scyllav1.RackSpec, scalingPolicy *v1alpha1.RackScalingPolicy) (*v1alpha1.RackRecommendations, error) {
+	var err error
+	var priority int32 = math.MaxInt32
+	members := rack.Members
+	resources := rack.Resources
+
+	applied := false
+	for _, rule := range scalingPolicy.ScalingRules {
+		if rule.Priority >= priority {
+			continue // TODO solve conflicting priorities, i.e. two rules with equal priorities???
+		}
+
+		var res bool
+		if rule.For != nil {
+			var step *time.Duration = nil
+			if rule.Step != nil {
+				step = &rule.Step.Duration
+			}
+			res, err = r.metricsProvider.RangedQuery(ctx, rule.Expression, rule.For.Duration, step)
+		} else {
+			res, err = r.metricsProvider.Query(ctx, rule.Expression)
+		}
+
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("rule \"%s\"", rule.Name))
+		}
+
+		if !res {
 			continue
 		}
+
+		if rule.ScalingMode == v1alpha1.ScalingModeHorizontal {
+			var min, max *int32 = nil, nil
+			if scalingPolicy.MemberPolicy != nil {
+				max = scalingPolicy.MemberPolicy.MaxAllowed
+				min = scalingPolicy.MemberPolicy.MinAllowed
+			}
+
+			members = calculateMembers(rack.Members, min, max, rule.ScalingFactor)
+		} else {
+			if rack.Resources.Requests == nil || rack.Resources.Requests.Cpu() == nil {
+				return nil, errors.New("cpu requests undefined")
```

Comment:
include info which rack doesn't have this info

Reply:
the error is wrapped higher in the call stack (errors.Wrapf(err, "rack \"%s\"", rack.Name))